command/helpers.go

```diff
@@ -0,0 +1,361 @@
+		if err = h.ItemCounts.RecordChange(*entity.DataType, false, savedInSQL); err != nil {
+			return false, err
+		}
+	}
+	return conflict, nil
+}
+
+func getMigratedEntityID(entity *datastore.SyncEntity) (string, error) {
+	id := entity.ID
+	if *entity.DataType == datastore.HistoryTypeID {
+		// In Dynamo, History entities are stored with the client tag as the ID.
+		// Since the SQL table uses a UUID for the id, generate a new ID here.
+		newID, err := uuid.NewV7()
+		if err != nil {
+			return "", err
+		}
+		id = newID.String()
+	}
+	return id, nil
+}
+
+func (h *DBHelpers) updateSyncEntity(entity *datastore.SyncEntity, oldVersion int64) (conflict bool, migratedEntity *datastore.SyncEntity, err error) {
+	var deleted bool
+	shouldSaveInSQL := h.SQLDB.Variations().ShouldSaveToSQL(*entity.DataType, h.variationHashDecimal)
+	if shouldSaveInSQL {
+		conflict, deleted, err = h.SQLDB.UpdateSyncEntity(h.Trx, entity, oldVersion)
+		if err != nil {
+			return false, nil, err
+		}
+		// Conflict might mean that the entity does not exist in SQL but exists in Dynamo.
+		// Check for a Dynamo entity and migrate it accordingly.
+		if conflict {
+			oldEntity, err := h.dynamoDB.GetEntity(datastore.ItemQuery{
+				ID:       entity.ID,
+				ClientID: entity.ClientID,
+			})
+			if err != nil {
+				return false, nil, err
+			}
+			if oldEntity == nil {
+				// The conflict is unrelated to a pending Dynamo to SQL migration.
+				// Return conflict error to client.
+				return true, nil, nil
+			}
+			if oldEntity.Deleted == nil || !*oldEntity.Deleted {
+				// If the stored entity was not already deleted, decrement the
+				// Dynamo item count since we'll be migrating the entity to SQL.
+				if err = h.ItemCounts.RecordChange(*entity.DataType, true, false); err != nil {
+					return false, nil, err
+				}
+			}
+			migratedEntityID, err := getMigratedEntityID(entity)
+			if err != nil {
+				return false, nil, err
+			}
+			entity.ID = migratedEntityID
+			conflict, err = h.SQLDB.InsertSyncEntities(h.Trx, []*datastore.SyncEntity{entity})
+			if err != nil {
+				return false, nil, err
+			}
+			if !conflict && (entity.Deleted == nil || !*entity.Deleted) {
+				// If the new entity is not considered deleted, increment the
+				// SQL interim count.
+				if err = h.ItemCounts.RecordChange(*entity.DataType, false, true); err != nil {
+					return false, nil, err
+				}
+			}
+			return conflict, oldEntity, err
+		}
+	} else {
+		conflict, deleted, err = h.dynamoDB.UpdateSyncEntity(entity, oldVersion)
+		if err != nil {
+			return false, nil, err
+		}
+	}
+	if !conflict && deleted {
+		if err = h.ItemCounts.RecordChange(*entity.DataType, true, shouldSaveInSQL); err != nil {
+			return false, nil, err
+		}
+	}
+	return conflict, nil, err
+}
+
+func (h *DBHelpers) maybeMigrateToSQL(dataTypes []int) (migratedEntities []*datastore.SyncEntity, err error) {
+	if !h.SQLDB.Variations().Ready {
+		return nil, nil
+	}
+	if rand.Float32() > h.SQLDB.MigrateIntervalPercent() {
```

Comment:
@DJAndries what's the purpose in this validation check? The logic is non-deterministic in the sense that rand could return a very large number forcing this to fail the migration. Is this intentional?

Reply:
to reduce DB load, we only want to migrate chunks of data from ddb to pg on a certain percentage of "get update" requests from the client, which are sent every 30 seconds. the check enforces this requirement.

Reply:
will add a comment for clarity

Reply:
If you want a more uniform randomness, crypto/rand can provide that guarantee, but based on that answer it doesn't seem like we really need that so feel free to ignore that sec-actions comment.